src/main/java/org/truffleruby/core/regexp/InterpolatedRegexpNode.java

```diff
@@ -9,71 +9,115 @@
  */
 package org.truffleruby.core.regexp;
 
-import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
-import com.oracle.truffle.api.frame.VirtualFrame;
-import com.oracle.truffle.api.nodes.ExplodeLoop;
-import com.oracle.truffle.api.object.DynamicObject;
 import org.jcodings.specific.ASCIIEncoding;
 import org.jcodings.specific.USASCIIEncoding;
 import org.truffleruby.Layouts;
+import org.truffleruby.Log;
+import org.truffleruby.core.regexp.InterpolatedRegexpNodeFactory.RegexpBuilderNodeGen;
 import org.truffleruby.core.rope.Rope;
 import org.truffleruby.core.rope.RopeBuilder;
+import org.truffleruby.core.rope.RopeNodes;
 import org.truffleruby.core.rope.RopeOperations;
 import org.truffleruby.core.string.StringOperations;
 import org.truffleruby.language.RubyNode;
 import org.truffleruby.language.dispatch.CallDispatchHeadNode;
 import org.truffleruby.parser.BodyTranslator;
 
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.NodeChild;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
+import com.oracle.truffle.api.object.DynamicObject;
+
 public class InterpolatedRegexpNode extends RubyNode {
 
     @Children private final RubyNode[] children;
-    private final RegexpOptions options;
     @Child private CallDispatchHeadNode toS = CallDispatchHeadNode.create();
+    @Child private RegexpBuilderNode builderNode;
 
     public InterpolatedRegexpNode(RubyNode[] children, RegexpOptions options) {
         this.children = children;
-        this.options = options;
+        builderNode = RegexpBuilderNode.create(options);
     }
 
     @Override
     public Object execute(VirtualFrame frame) {
-        return createRegexp(executeChildren(frame));
+        return builderNode.execute(executeChildren(frame));
     }
 
-    @TruffleBoundary
-    private DynamicObject createRegexp(DynamicObject[] parts) {
-        final Rope[] strings = new Rope[children.length];
+    @ExplodeLoop
```

Comment:
Out of curiosity, what's the largest number of children you've encountered? These `@ExplodeLoop` with no upper bound always concern me.

Reply:
I think the most I've seen is 4, but it's not the easiest thing to grep for in source because the the source can be multi-line, have unusual delimiters and so forth, so I haven't done a large scale check. Want me to put in a guard for an upper bound?

Reply:
This would have to be a source literal to be interpolated, right? It seems unlikely anyone would push that more than say, 20. An upper-bound may not be practically necessary.

Reply:
Yes, these things are only from source literals so the number is usually quite small and it never varies for an individual site.

Reply:
It seems wise to have a limit, e.g. generated code could go above these limits.

Reply:
Although this seems unlikely to ever be too large, we'd have the same issue with a very large number of statements inside a single method, etc.